# Hacker news project

This project concerns the data of the hackernews.com website. We are going to analyse what type of post are getting more comments than others. In order to do that we will compute the best type of post and the best time to post. 

In [1]:
from csv import reader
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)
print(hn[0:5])


[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [2]:
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[0:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [3]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


In [6]:
total_ask_comments = 0
for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments
    
avg_ask_comments = total_ask_comments / 1744
print(avg_ask_comments)

total_show_comments = 0
for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments

avg_show_comments = total_show_comments / 1162
print(avg_show_comments)    
    

14.038417431192661
10.31669535283993


## Ask versus show comments

For ask comments the average is 14.03
While for show comments it is 10.31
This difference in average is probably due to the fact that people on HackerNews are more ready to respond to questions than to things that are just being shown. 

In [14]:
import datetime as dt
result_list = []
for row in ask_posts:
    created_at = row[6]
    n_comments = int(row[4])
    result_list.append([created_at, n_comments])

counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    t_post = dt.datetime.strptime(row[0], "%m/%d/%Y %H:%M")
    h_post = t_post.strftime("%H")
    if h_post not in counts_by_hour:
        counts_by_hour[h_post] = 1
        comments_by_hour[h_post] = row[1]
    else:
        counts_by_hour[h_post] += 1
        comments_by_hour[h_post] += row[1]
        
print(counts_by_hour)
        
    

{'04': 47, '16': 108, '21': 109, '03': 54, '09': 45, '22': 71, '17': 100, '08': 48, '18': 109, '20': 80, '06': 44, '00': 55, '14': 107, '23': 68, '02': 58, '01': 60, '15': 116, '19': 110, '11': 58, '12': 73, '10': 59, '07': 34, '13': 85, '05': 46}


In [17]:
avg_by_hour = []

for hour in comments_by_hour:
    avg_by_hour.append([hour, (comments_by_hour[hour]/counts_by_hour[hour])])
                        
print(avg_by_hour)

[['04', 7.170212765957447], ['16', 16.796296296296298], ['21', 16.009174311926607], ['03', 7.796296296296297], ['09', 5.5777777777777775], ['22', 6.746478873239437], ['17', 11.46], ['08', 10.25], ['18', 13.20183486238532], ['20', 21.525], ['06', 9.022727272727273], ['00', 8.127272727272727], ['14', 13.233644859813085], ['23', 7.985294117647059], ['02', 23.810344827586206], ['01', 11.383333333333333], ['15', 38.5948275862069], ['19', 10.8], ['11', 11.051724137931034], ['12', 9.41095890410959], ['10', 13.440677966101696], ['07', 7.852941176470588], ['13', 14.741176470588234], ['05', 10.08695652173913]]


In [22]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
print(swap_avg_by_hour)

[[7.170212765957447, '04'], [16.796296296296298, '16'], [16.009174311926607, '21'], [7.796296296296297, '03'], [5.5777777777777775, '09'], [6.746478873239437, '22'], [11.46, '17'], [10.25, '08'], [13.20183486238532, '18'], [21.525, '20'], [9.022727272727273, '06'], [8.127272727272727, '00'], [13.233644859813085, '14'], [7.985294117647059, '23'], [23.810344827586206, '02'], [11.383333333333333, '01'], [38.5948275862069, '15'], [10.8, '19'], [11.051724137931034, '11'], [9.41095890410959, '12'], [13.440677966101696, '10'], [7.852941176470588, '07'], [14.741176470588234, '13'], [10.08695652173913, '05']]


In [30]:
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print("Top 5 Hours for Ask Posts Comments")
for row in sorted_swap[0:5]:
    string = "{}:00: {:.2f} average comments per post"
    print(string.format(dt.datetime.strptime(row[1], '%H').strftime("%H"), row[0]))

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


## Results

The last step in determining the difference in time zone between the data and the current time zone we are in. If we check the following [link](https://www.kaggle.com/hacker-news/hacker-news-posts), the times for the dataset are in Eastern time US.

The difference between Eastern time US and GMT +1 (Amsterdam) is five hours. In the US it is five hours earlier than in Amsterdam. 

This gives us a final result:

|Time |Average C/P|
|---|---|
|20:00   | 38.59 |
|7:00   | 23.81  |
|1:00   | 21.52  |
|21:00   | 16.80  |
|2:00   | 16.01  |

The best hour is clearly 20:00. But if you miss that window, 21:00 is still a nice second best. Unless you would like to get up early or stay awake past midnight. 